Assignments/HW2/hw2.md

```diff
@@ -0,0 +1,173 @@
+
+### Task 1.3: Include current conversations in join response [10 points]
+If you completed task 1.2 correctly, then all players who are in the town when a new converation is made will be notified of the new converation area. The last task to complete before we can move on to frontend implementation of this feature is to update the backend service to include the list of all current conversation areas when a new player joins the town. 
+
+In the file `CoveyTownRequestHandler.ts`, examine the method `townJoinHandler` and the type `TownJoinResponse`. Your objective for this task is to update the `TownJoinResponse` type to include a new field, `conversationAreas`, of type `ServerConversationArea[]`, and to update the `townJoinHandler` to include the current list of converation areas for the town that the user has joined.
+
+📝 Check your work: Run `npm test CoveyTownConversationAPI` again, you should now see the test "Includes newly created conversations when a new player joins" pass.
+
+### Checkpoint, check your work on this entire task
+We suggest that you submit your code at this point on GradeScope, which will run Avery's *extremely* thorough test suite on your code. If there are test errors, you will likely find it easier to debug them *now* rather than moving on to implement more code (and perhaps introduce more bugs).
+
+## Part 2: Completing the TownService responsibilities [30 points total]
+Congratulations on making it this far! You are almost done, there are only two more tasks to complete the feature.
+
+### Task 2.1: Track conversation participants [15 points]
+In Avery's design, each of the users connected to a town track which conversation area (if any) they are in, and send this information to the `CoveyTownController`. The `CoveyTownController` needs to track which users are in each conversation. `CoveyTownController` has a method `updatePlayerLocation`, which is called each time that a player's location changes. Your objective for this task is to implement functionality so that at the end of the execution of this method, the `_conversationAreas` list tracked by the town controller reflects that player's movement (updating the `occupantsByID` property on any effected conversation areas). If any conversation areas are updated, you must emit a `onConversationAreaUpdated` event, similarly to how you did in Task 1.2 when a conversation is created.
+
+Note that you might need to send multiple `onConversationAreaUpdated` events: one for a user leaving an area and one for them entering another. It is important that no client ever believe that a user is in two conversation areas at the same time. Hence, be sure to send the "exit" update before the "enter" update.
+
+📝 Check your work: Again, Avery has provided a single sanity test that checks some of the behavior defined above. You can run this test with the command `npm test -- -t 'CoveyTownController updatePlayerLocation'`, and you might also consider extending it.
```

Comment:
This test fails for me after I wrote the logic for the `updatePlayerLocation()` method. After going through the test, I feel there might be a typo in the test case, as the conversation area bounding box coordinates are `(100, 100)` with width and height of 20 each. However, the player's location is `(10, 10)`, which is far away from the conversation area, but the test expects the player's conversation area to be updated.

Reply:
Good catch, thank you! I had fixed this for the gradescope tests, but had forgotten to propagate it back to the handout test. I'm working on getting the gradescope script set up right now.

Reply:
> Good catch, thank you! I had fixed this for the gradescope tests, but had forgotten to propagate it back to the handout test. I'm working on getting the gradescope script set up right now.

Sure professor! Let me know if you need help testing the same.